src/Microsoft.AspNetCore.Server.Kestrel/Http/Connection.cs

```diff
@@ -252,8 +252,16 @@ private static void ReadCallback(UvStreamHandle handle, int status, object state
 
         private void OnRead(UvStreamHandle handle, int status)
         {
+            if (status == 0)
+            {
+                // A zero status does not indicate an error or connection end. It indicates
+                // there is no data to read right now and can be ignored.
+                // See the note at http://docs.libuv.org/en/v1.x/stream.html#c.uv_read_cb.
+                return;
```

Comment:
So I was wrong about being able to simply return here.

The reason libuv is calling our read callback without any data is to allow us to cleanup after our alloc callback. This explains the Debug.Assert failure we saw yesterday.

Since our alloc callback calls `SocketInput.IncomingStart`, we need to null out `SocketInput._pinned`. Before that, if `_pinned != _tail`, we should return `_pinned` to the pool.

I would just add another method to `SocketInput` that does this. Maybe call it `IncomingDeferred`.